### ISSUE TITLE
dbWriteTable won't write into tables with autoicrement (INTEGER PRIMARY KEY) fields

### ISSUE THREAD

**Issue by rickyars** [rstats-db/RSQLite#59]
Prior to the update, my code was working just fine. I set the row.names=FALSE and all was good. Today I ran my code and I'm getting a message saying that I'm one column short. Did something change?

**Comment by hadley** [comment]
This seems to be fixed now:

```R
# connect to the database
db <- dbConnect(SQLite())
dbGetQuery(db, "CREATE TABLE Test
(
  TestId INTEGER PRIMARY KEY, -- Autoincrement
  TestVariable TEXT,
  TestValue INTEGER
)")

# made up data
test.df <- data.frame(TestVariable = letters[sample.int(26, 10)], TestValue = runif(10))

# write data frame to database
dbWriteTable(db, "Test", test.df, append=TRUE, row.names=FALSE)
```

**Comment by rickyars** [comment]
I updated all my packages but I'm still having this problem.

**Comment by hadley** [comment]
Did you install the dev version?

**Comment by rickyars** [comment]
No. I'll give that a shot.

**Comment by MaxStreese** [comment]
Hi, do I still need to install the dev version for this to work? I am getting the same exact error message.

**Comment by Steviey** [comment]
+1 ...poor quality

**Comment by MaxStreese** [comment]
For anyone still having this problem: When constructing the dataframe just include the id column but set it to NA. That works for me.

**Comment by Steviey** [comment]
@MaxStreese Does then the function leaves a primary key of id in a db-table unchanged?

**Comment by hadley** [comment]
If you're having problems, please file a new issue with a reprex. Commenting on a three year old issue is not a productive way to get help.

**Comment by MaxStreese** [comment]
@Steviey not sure what you mean. When you ask what I think you are asking then: No. If you call 

`dbWriteTable(someConnection, "SomeTable", some.data.frame, append=TRUE, row.names=FALSE)`

On  a table that has an `INTEGER PRIMARY KEY` as the first column and your `some.data.frame` is filled with `NA` in the first column then your data will get inserted into the database and the ID will be incremented for each record.

**Comment by hadley** [comment]
There is no need to use `NA`:

``` r
library(DBI)

db <- dbConnect(RSQLite::SQLite(), filename = ":memory:")
dbExecute(db, "CREATE TABLE test
(
  x INTEGER PRIMARY KEY, -- Autoincrement
  y INTEGER
)")
#> [1] 0

dbWriteTable(db, "test", data.frame(y = 5:1), append = TRUE)
dbReadTable(db, "test")
#>   x y
#> 1 1 5
#> 2 2 4
#> 3 3 3
#> 4 4 2
#> 5 5 1
```

Created on 2018-03-01 by the [reprex package
(0.1.1.9000)](https://cran.r-project.org/package=reprex).

**Comment by Steviey** [comment]
I misinterpreted another sql error. Sorry, seems to work.

**Comment by thompsonlaurak** [comment]
I'm having this issue but with SQL Server, not SQL Lite. None of the above solutions work for me. I have used @hadley 's reprex and provided my error messages. I'm new to GitHub (and SQL for that matter, but not R), so apologies if I'm not following standard protocol here...Should I make a new issue?

library(DBI)

con <- dbConnect(odbc::odbc(), Driver = "SQL Server", Server = "YourServerHere", 
            Database = "YourDatabaseHere", ...)

dbExecute(con, "CREATE TABLE test
(
  x INTEGER PRIMARY KEY, -- Autoincrement
  y INTEGER
)")

_#> [1] 0_

dbWriteTable(con, "test", data.frame(y = 5:1), append = TRUE)

_#> Error in result_insert_dataframe(rs@ptr, values) :    nanodbc/nanodbc.cpp:1587: 23000: [Microsoft][ODBC SQL Server Driver][SQL Server]Cannot insert the value NULL into column 'x', table 'Redacted.dbo.test'; column does not allow nulls. INSERT fails._

I also tried replacing data.frame(y = 5:1) with data.frame(x = NA, y = 5:1) since a few people said filling the primary key column with NA's  worked. That results in this error message: 

_#> Error in result_insert_dataframe(rs@ptr, values) : 
  nanodbc/nanodbc.cpp:1587: 23000: [Microsoft][ODBC SQL Server Driver][SQL Server]Cannot insert the value NULL into column 'x', table 'Redacted.dbo.test'; column does not allow nulls. INSERT fails._

Is there something fundamentally different about SQL Server that is preventing me from using either of these solutions?
